### PR TITLE
fix(datasource/go): re-arrange the go datasource extraction in functions 

### PR DESCRIPTION
**Pull request by PhilipAbed**
## Changes

changed the behavior, Instead of "if/else", its now Try first, if failed, try second.

## Context

related to issue #15333 but will not close it. just a refactor before the fix.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Is it "changing the behavior" as in your description, or just a pure refactor with no behavior change?

Also please use a semantic PR title

**Comment by PhilipAbed** [comment]
i changed from 
```
If(go-source found) do parse it.
else try to find go-import and parse it 
```
to 
`try to find go-source and parse it ?? try to find go import and parse it`

the behavior change is just an edge case scenario where go-source is found but failed to parse, now this shouldnt happen usually but in this case, we will try the go-import now...

everything else is the same still.

**Comment by PhilipAbed** [comment]
should i change from refactor to something else?

File: lib/modules/datasource/go/base.ts [hunk 1]
```diff
@@ -53,120 +53,137 @@ export class BaseGoDatasource {
...
-          packageName,
-        };
+  private static goImportHeader(
+    res: string,
+    goModule: string
+  ): DataSource | null {
+    const importMatch = regEx(
+      `<meta\\s+name="go-import"\\s+content="([^\\s]+)\\s+([^\\s]+)\\s+([^\\s]+)">`
+    ).exec(res);
+    if (importMatch) {
```
Comment (rarkins) [inline lib/modules/datasource/go/base.ts:160]: I think switch this to negative:

```suggestion
    if (!importMatch) {
```

And then return early.

**Review by rarkins — CHANGES_REQUESTED** [review]
If behavior changes then it's a fix not a refactor

File: lib/modules/datasource/go/base.ts (hunk 1 above)
Comment (PhilipAbed) [reply lib/modules/datasource/go/base.ts:160]: didnt even notice that xD yes sure.

**Review by rarkins — APPROVED** [review]

File: lib/modules/datasource/go/base.ts [hunk 2]
```diff
@@ -53,120 +53,138 @@ export class BaseGoDatasource {
...
 
-        // TODO: `parsedUrl.pathname` can be undefined
-        const packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
+    const opts = hostRules.find({
+      hostType: PlatformId.Gitlab,
+      url: goSourceUrl,
+    });
+    if (opts.token) {
+      // get server base url from import url
+      const parsedUrl = URL.parse(goSourceUrl);
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:136]: shouldn't we use our url utils here? or is it copy and paste?

File: lib/modules/datasource/go/base.ts [hunk 3]
```diff
@@ -53,120 +53,138 @@ export class BaseGoDatasource {
...
+
+    const [, prefix, , goImportURL] = importMatch;
+    if (!goModule.startsWith(prefix)) {
+      logger.trace({ goModule }, 'go-import header prefix not match');
+      return null;
+    }
+
+    logger.debug({ goModule, goImportURL }, 'Go lookup import url');
+    // get server base url from import url
+    const parsedUrl = URL.parse(goImportURL);
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:174]: same here

File: lib/modules/datasource/go/base.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/datasource/go/base.ts:136]: This is a copy/paste from the old, I confirmed by hiding white space when looking at the diff. Same as below

File: lib/modules/datasource/go/base.ts (hunk 3 above)
Comment (PhilipAbed) [reply lib/modules/datasource/go/base.ts:174]: copy paste i didnt touch

**Pull request merged**